File: src/main/java/org/modelingvalue/nelumbo/integers/Multiply.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.integers;

import java.math.BigInteger;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.InferContext;
import org.modelingvalue.nelumbo.InferResult;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.logic.Predicate;
import org.modelingvalue.nelumbo.patterns.Functor;

public final class Multiply extends Predicate {
    private static final long serialVersionUID = 2630128775301942610L;

    public Multiply(Functor functor, List<AstElement> elements, Object[] args) {
        super(functor, elements, args[0], args[1], args[2]);
    }

    private Multiply(Object[] array, Multiply declaration) {
        super(array, declaration);
    }

    @Override
    protected Multiply struct(Object[] array, Node declaration) {
        return new Multiply(array, (Multiply) declaration);
    }

    @Override
    protected InferResult infer(int nrOfUnbound, InferContext context) {
        if (nrOfUnbound > 1) {
            return unresolvable();
        }

        BigInteger factor1 = getVal(0, 0);
        BigInteger factor2 = getVal(1, 0);
        BigInteger product = getVal(2, 0);
        if (factor1 != null && factor2 != null) {
            BigInteger p = factor1.multiply(factor2);
            if (product != null) {
                boolean eq = p.equals(product);
                return eq ? factCC() : falsehoodCC();
            } else {
                return set(2, NInteger.of(p)).factCI();
            }
        } else if (factor1 != null && product != null) {
            BigInteger[] dr = product.divideAndRemainder(factor1);
            return dr[1].equals(BigInteger.ZERO) ? set(1, NInteger.of(dr[0])).factCI() : falsehoodCI();
        } else if (factor2 != null && product != null) {
            BigInteger[] dr = product.divideAndRemainder(factor2);
            return dr[1].equals(BigInteger.ZERO) ? set(0, NInteger.of(dr[0])).factCI() : falsehoodCI();
        }

        return unknown();
    }

}

File: web-ts/src/integers/Multiply.ts
/**
 * Multiply - Integer multiplication predicate.
 * @JAVA_REF org.modelingvalue.nelumbo.integers.Multiply
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Predicate } from '../logic/Predicate';
import type { Functor } from '../patterns/Functor';
import type { InferContext } from '../InferContext';
import type { InferResult } from '../InferResult';
import { NInteger } from './NInteger';

export class Multiply extends Predicate {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, args[0], args[1], args[2]);
  }

  protected static fromDataMultiply(data: unknown[], declaration?: Predicate): Multiply {
    const node = Object.create(Multiply.prototype) as Multiply;
    (node as unknown as { _data: unknown[] })._data = data;
    (node as unknown as { _declaration: Predicate | undefined })._declaration = declaration ?? node;
    return node;
  }

  protected override struct(data: unknown[], declaration?: Predicate): Predicate {
    return Multiply.fromDataMultiply(data, declaration ?? this.declaration());
  }

  override infer(_context: InferContext): InferResult {
    const nrOfUnbound = this.nrOfUnbound();
    if (nrOfUnbound > 1) {
      return this.unresolvable();
    }

    const factor1 = this.getIntegerVal(0);
    const factor2 = this.getIntegerVal(1);
    const product = this.getIntegerVal(2);

    if (factor1 !== null && factor2 !== null) {
      const p = factor1 * factor2;
      if (product !== null) {
        const eq = p === product;
        return eq ? this.factCC() : this.falsehoodCC();
      } else {
        return (this.set(2, NInteger.of(p)) as Multiply).factCI();
      }
    } else if (factor1 !== null && product !== null && factor1 !== 0n) {
      const quotient = product / factor1;
      const remainder = product % factor1;
      return remainder === 0n
        ? (this.set(1, NInteger.of(quotient)) as Multiply).factCI()
        : this.falsehoodCI();
    } else if (factor2 !== null && product !== null && factor2 !== 0n) {
      const quotient = product / factor2;
      const remainder = product % factor2;
      return remainder === 0n
        ? (this.set(0, NInteger.of(quotient)) as Multiply).factCI()
        : this.falsehoodCI();
    }

    return this.unknown();
  }

  private getIntegerVal(index: number): bigint | null {
    const val = this.get(index);
    if (val instanceof NInteger) {
      return val.value();
    }
    return null;
  }
}
